### PR TITLE
Unit tests for Utils.concat() 

### LINKED ISSUE #7916: The Utils.concat() methods need unit tests
The various overloads of `Utils.concat()` are not covered by unit tests.

### PR DESCRIPTION
**Pull request by orlicohen**
Fixes #7916. Data providers and unit tests for the various overloads of Utils.concat()

### CODE REVIEW

File: src/test/java/org/broadinstitute/hellbender/utils/UtilsUnitTest.java [hunk 1]
```diff
@@ -978,4 +981,78 @@ public void testTestFilterCollectionByExpressions(Set<String> values, Collection
...
+    @DataProvider(name = "concatMultipleByteArraysData")
+    public Object[][] concatMultipleByteArraysData() {
+        return new Object[][]{
+                // expected, arrays
+                new Object[]{ ArrayUtils.EMPTY_BYTE_ARRAY, new byte[] { } },
+                new Object[]{ new byte[] {100, 100}, new byte[] {100, 100} },
+                new Object[]{ new byte[] {100, 100}, new byte[] {100, 100}, new byte[] { } },
+                new Object[]{ ArrayUtils.EMPTY_BYTE_ARRAY, new byte[] { }, new byte[] { } },
+                new Object[]{ new byte[] {100, 100, 100, 100}, new byte[] {100, 100}, new byte[] {100, 100} },
+        };
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/UtilsUnitTest.java:994]: Comments on the test cases for the varargs version of concat():

- Calling concat() with 2 byte arrays actually calls the two-arg specialization of concat(), not the varargs version, so none of your test cases here should have 2 input arrays. Call concat() with 0, 1, or 3+ arrays to test the varargs version.
- The values inside your arrays should not all be the same (100), otherwise you are not actually testing whether the method is preserving the original order of the values correctly during concatenation.

File: src/test/java/org/broadinstitute/hellbender/utils/UtilsUnitTest.java [hunk 2]
```diff
@@ -978,4 +981,78 @@ public void testTestFilterCollectionByExpressions(Set<String> values, Collection
...
+    }
+
+    @DataProvider(name = "concatTwoByteArraysData")
+    public Object[][] concatTwoByteArraysData() {
+        return new Object[][]{
+                // arr1, arr2, expected
+                new Object[]{ new byte[] { }, new byte[] { }, ArrayUtils.EMPTY_BYTE_ARRAY },
+                new Object[]{ new byte[] {100, 100}, new byte[] { }, new byte[] {100, 100} },
+                new Object[]{ new byte[] { }, new byte[] {100, 100}, new byte[] {100, 100} },
+                new Object[]{ new byte[] {100, 100}, new byte[] {100, 100}, new byte[] {100, 100, 100, 100} },
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/UtilsUnitTest.java:1015]: Comments on the test cases for the two-arg version of concat():

- As above, the values inside your arrays should not all be the same (100), otherwise you are not actually testing whether the method is preserving the original order of the values correctly during concatenation.
- Include at least one additional test case involving an array of length 1, as that is a common edge case for methods like this.

File: src/test/java/org/broadinstitute/hellbender/utils/UtilsUnitTest.java [hunk 3]
```diff
@@ -978,4 +981,78 @@ public void testTestFilterCollectionByExpressions(Set<String> values, Collection
...
+                // arr1, arr2, constructor, expected
+                new Object[]{ new Integer[] { }, new Integer[] { }, (IntFunction<Integer[]>)Integer[]::new,
+                        new Integer[] { } },
+                new Object[]{ new Integer[] {1, 2, 3}, new Integer[] { }, (IntFunction<Integer[]>)Integer[]::new,
+                        new Integer[] {1, 2, 3} },
+                new Object[]{ new Integer[] { }, new Integer[] {4, 5, 6}, (IntFunction<Integer[]>)Integer[]::new,
+                        new Integer[] {4, 5, 6} },
+                new Object[]{ new Integer[] {1, 2, 3}, new Integer[] {4, 5, 6}, (IntFunction<Integer[]>)Integer[]::new,
+                        new Integer[] {1, 2, 3, 4, 5, 6} },
+        };
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/UtilsUnitTest.java:1036]: Comments on the test cases for the generic version of concat():

- Include an additional test case with { 4, 5, 6 } and { 1, 2, 3 } passed in the reverse order, giving { 4, 5, 6, 1, 2, 3 }. This will demonstrate that the method is preserving input order correctly and not doing something silly like sorting the values during concatenation.

File: src/test/java/org/broadinstitute/hellbender/utils/UtilsUnitTest.java [hunk 4]
```diff
@@ -978,4 +981,78 @@ public void testTestFilterCollectionByExpressions(Set<String> values, Collection
...
+       return new Object[][]{
+               new Object[]{null, new Integer[]{1, 2, 3}, (IntFunction<Integer[]>) Integer[]::new},
+               new Object[]{new Integer[] {1, 2, 3}, null, (IntFunction<Integer[]>) Integer[]::new},
+               new Object[]{null, null, (IntFunction<Integer[]>) Integer[]::new}
+       };
+   }
+
+    @Test(expectedExceptions = IllegalArgumentException.class, dataProvider = "concatAnyTypeWithNullArrayData")
+    public void testConcatAnyTypeWithNullArray(Integer[] arr1, Integer[] arr2, IntFunction<Integer[]> constructor){
+        Utils.concat(arr1, arr2, constructor);
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/UtilsUnitTest.java:1055]: Null tests look good 👍

File: src/test/java/org/broadinstitute/hellbender/utils/UtilsUnitTest.java [hunk 5]
```diff
@@ -28,7 +30,8 @@
  * Testing framework for general purpose utilities class.
  *
  */
-public final class UtilsUnitTest extends GATKBaseTest {
+public final class
+UtilsUnitTest extends GATKBaseTest {
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/UtilsUnitTest.java:34]: Class name should be on the same line as the declaration

**Review by droazen — CHANGES_REQUESTED** [review]
@orlicohen Looks good, just a few minor changes needed

**Review by droazen — CHANGES_REQUESTED** [review]
@orlicohen Two quick remaining comments, then this should be good to go

**Review by droazen — APPROVED** [review]
